machine/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlconfigv1 "sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const (
+	ProviderFinalizer         = "provider.cluster.x-k8s.io"
+	ConfigMapVersionLabelName = "provider.cluster.x-k8s.io/version"
+)
+
+// ProviderSpec is the desired state of the Provider.
+type ProviderSpec struct {
+	// Version indicates the provider version.
+	Version string `json:"version,omitempty"`
+
+	// Manager defines the properties that can be enabled on the controller manager for the provider.
+	// +optional
+	Manager *ManagerSpec `json:"manager,omitempty"`
+
+	// Deployment defines the properties that can be enabled on the deployment for the provider.
+	// +optional
+	Deployment *DeploymentSpec `json:"deployment,omitempty"`
+
+	// SecretName is the name of the Secret providing the configuration
+	// variables for the current provider instance, like e.g. credentials.
+	// Such configurations will be used when creating or upgrading provider components.
+	// The contents of the secret will be treated as immutable. If changes need
+	// to be made, a new object can be created and the name should be updated.
+	// The contents should be in the form of key:value. This secret must be in
+	// the same namespace as the provider.
+	// +optional
+	SecretName *string `json:"secretName,omitempty"`
+
+	// FetchConfig determines how the operator will fetch the components and metadata for the provider.
+	// If nil, the operator will try to fetch components according to default
+	// embedded fetch configuration for the given kind and `ObjectMeta.Name`.
+	// For example, the infrastructure name `aws` will fetch artifacts from
+	// https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases.
+	// +optional
+	FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`
+}
+
+// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.
+type ManagerSpec struct {
+	// ControllerManagerConfigurationSpec defines the desired state of GenericControllerManagerConfiguration.
+	ctrlconfigv1.ControllerManagerConfigurationSpec `json:",inline"`
```

Comment:
dependencies on other groupversions are discouraged because they tend to fall out of date

Reply:
And I have no idea what fields are in this struct.

Reply:
I will extract the required fields from ControllerManagerConfigurationSpec to it obvious.

Reply:
Just as a point of note, this pattern is [recommended as part of the kubebuilder project](https://book.kubebuilder.io/component-config-tutorial/config-type.html) for extending configuration parameters, if there are good reasons that this becomes problematic, perhaps we should gently nudge that community not to do this or at least add some warning to their docs? (I was considering doing this for a new project soon)